File: js/input.js
import { HUD_H, MAP_PIXEL_H, MAP_PIXEL_W, SHOP_H, TOWER_TYPES } from './config.js';
import { isNobuild, isWall } from './map.js';
import { pathExistsWithTowerAt } from './pathfinder.js';
import { Tower } from './entities.js';
import { setupHeli } from './combat.js';
import { Phase } from './state.js';
import { posToCell } from './map.js';

/** Canvas coords → map surface coords (strip HUD strip above grid). */
export function canvasToMap(mx, my) {
  const mapY = my - HUD_H;
  if (mx < 0 || mx >= MAP_PIXEL_W || mapY < 0 || mapY >= MAP_PIXEL_H) return null;
  return [mx, mapY];
}

export function screenToGrid(mx, my) {
  const map = canvasToMap(mx, my);
  if (!map) return null;
  return posToCell(map[0], map[1]);
}

/** Tower id 1–9 if click is on the shop bar. */
export function shopTowerAt(mx, my) {
  const y0 = HUD_H + MAP_PIXEL_H;
  if (my < y0 || my >= y0 + SHOP_H || mx < 0 || mx >= MAP_PIXEL_W) return null;
  const slotW = MAP_PIXEL_W / 9;
  const id = Math.floor(mx / slotW) + 1;
  return id >= 1 && id <= 9 ? id : null;
}

export function towerAtScreen(state, mx, my) {
  const cell = screenToGrid(mx, my);
  return cell ? state.towerAt(cell[0], cell[1]) : null;
}

export function canPlace(state, col, row) {
  if (isWall(col, row) || isNobuild(col, row)) return false;
  if (state.towerAt(col, row)) return false;
  return pathExistsWithTowerAt(col, row, state.towerCells);
}

export function updateHover(state, mx, my) {
  state.hoverCell = screenToGrid(mx, my);
  state.hoverTower = state.hoverCell ? state.towerAt(state.hoverCell[0], state.hoverCell[1]) : null;
}

function repathAll(state) {
  state.recomputePath();
  for (const e of state.enemies) e.requestRepath(state.globalPath);
}

export function tryPlace(state, col, row) {
  const tid = state.shopSelection;
  if (!tid || !TOWER_TYPES[tid]) return;
  if (!canPlace(state, col, row)) return;
  const cost = TOWER_TYPES[tid].base_cost;
  if (state.gold < cost) return;
  state.gold -= cost;
  const tower = new Tower(tid, col, row, cost);
  state.towers.push(tower);
  state.towerCells.add(`${col},${row}`);
  repathAll(state);
  if (tid === 9) {
    state.phase = Phase.HELI_SETUP;
    state.heliSetupTower = tower;
  } else {
    state.shopSelection = null;
  }
}

export function tryUpgrade(state) {
  const t = state.placedSelection;
  if (!t || !t.canUpgrade()) return;
  const cost = t.getUpgradeCost();
  if (state.gold < cost) return;
  state.gold -= cost;
  t.upgrade();
}

export function trySell(state) {
  const t = state.placedSelection;
  if (!t) return;
  state.gold += t.sellValue();
  state.towers = state.towers.filter(x => x.id !== t.id);
  state.towerCells.delete(`${t.col},${t.row}`);
  state.placedSelection = null;
  repathAll(state);
}

export function handleKey(state, waveMgr, key) {
  if (state.phase === Phase.GAME_OVER) {
    if (key === 'Enter' || key === 'r') state.resetRun();
    return;
  }
  if (state.phase === Phase.PAUSED) {
    if (key === 'Escape') state.phase = Phase.PLAYING;
    return;
  }
  if (state.phase !== Phase.PLAYING && state.phase !== Phase.HELI_SETUP) return;

  if (key === 'Escape') {
    if (state.phase === Phase.HELI_SETUP) {
      state.phase = Phase.PLAYING;
      state.heliSetupTower = null;
    } else state.phase = Phase.PAUSED;
    return;
  }
  if (key === '0') {
    state.shopSelection = null;
    state.placedSelection = null;
    return;
  }
  if (key >= '1' && key <= '9') {
    state.shopSelection = parseInt(key, 10);
    state.placedSelection = null;
    return;
  }
  if (key === ' ' && state.waveReady && !state.waveActive) {
    state.waveIndex++;
    state.recomputePath();
    state.waveActive = true;
    state.waveReady = false;
    waveMgr.startWave(state.waveIndex);
    return;
  }
  if (key === 'u' || key === 'U') tryUpgrade(state);
  if (key === 's' || key === 'S') trySell(state);
}

export function handleClick(state, mx, my, button) {
  if (state.phase === Phase.GAME_OVER) {
    if (button === 0 && mx > 400 && mx < 700 && my > 300 && my < 360) state.resetRun();
    return;
  }
  if (state.phase === Phase.PAUSED) {
    if (button !== 0) return;
    if (my > 200 && my < 250) state.phase = Phase.PLAYING;
    else if (my > 270 && my < 320) state.resetRun();
    return;
  }
  if (state.phase === Phase.HELI_SETUP && button === 0 && state.heliSetupTower) {
    const map = canvasToMap(mx, my);
    if (!map) return;
    const cx = Math.max(40, Math.min(map[0], MAP_PIXEL_W - 40));
    const cy = Math.max(40, Math.min(map[1], MAP_PIXEL_H - 40));
    const pad = state.heliSetupTower;
    const radius = Math.hypot(cx - pad.pos[0], cy - pad.pos[1]);
    setupHeli(pad, [cx, cy], radius);
    state.phase = Phase.PLAYING;
    state.heliSetupTower = null;
    state.shopSelection = null;
    return;
  }
  if (state.phase !== Phase.PLAYING) return;

  if (button === 0) {
    const shopId = shopTowerAt(mx, my);
    if (shopId) {
      state.shopSelection = shopId;
      state.placedSelection = null;
      return;
    }
    if (state.shopSelection) {
      const cell = screenToGrid(mx, my);
      if (cell) tryPlace(state, cell[0], cell[1]);
    }
    return;
  }

  if (button === 2) {
    const t = towerAtScreen(state, mx, my);
    state.placedSelection = t;
  }
}

File: game/input.py
"""Input handling."""

from __future__ import annotations

import pygame

from game.config import MAP_PIXEL_H, MAP_PIXEL_W, TOWER_TYPES
from game.map_data import is_nobuild, is_wall, pos_to_cell
from game.pathfinder import path_exists_with_tower_at
from game.state import GamePhase


def screen_to_grid(mx, my) -> tuple[int, int] | None:
    if mx >= MAP_PIXEL_W or my >= MAP_PIXEL_H:
        return None
    return pos_to_cell(mx, my)


def tower_at_screen(state, mx, my):
    cell = screen_to_grid(mx, my)
    if cell:
        return state.tower_at(cell[0], cell[1])
    return None


def can_place(state, col, row) -> bool:
    if is_wall(col, row) or is_nobuild(col, row):
        return False
    if state.tower_at(col, row):
        return False
    if not path_exists_with_tower_at(col, row, state.tower_cells):
        return False
    return True


def handle_keydown(key, state, wave_mgr, audio) -> None:
    if state.phase == GamePhase.MENU:
        if key in (pygame.K_RETURN, pygame.K_SPACE):
            state.reset_run()
            if audio:
                audio.play("wave")
        return

    if state.phase == GamePhase.GAME_OVER:
        if key in (pygame.K_RETURN, pygame.K_r):
            state.reset_run()
        return

    if state.phase == GamePhase.PAUSED:
        if key == pygame.K_ESCAPE:
            state.phase = GamePhase.PLAYING
        return

    if state.phase not in (GamePhase.PLAYING, GamePhase.HELI_SETUP):
        return

    if key == pygame.K_ESCAPE:
        if state.phase == GamePhase.HELI_SETUP:
            state.phase = GamePhase.PLAYING
            state.heli_setup_tower = None
        else:
            state.phase = GamePhase.PAUSED
        return

    if key == pygame.K_0:
        state.shop_selection = None
        state.placed_selection = None
        return

    if pygame.K_1 <= key <= pygame.K_9:
        state.shop_selection = key - pygame.K_0
        state.placed_selection = None
        return

    if key == pygame.K_SPACE and state.wave_ready and not state.wave_active:
        state.wave_index += 1
        state.wave_active = True
        state.wave_ready = False
        wave_mgr.start_wave(state.wave_index)
        if audio:
            audio.play("wave")
        return

    if key == pygame.K_u and state.placed_selection:
        _try_upgrade(state, audio)
    if key == pygame.K_s and state.placed_selection:
        _try_sell(state)


def handle_mouse(state, event, audio) -> None:
    mx, my = event.pos

    if state.phase == GamePhase.MENU:
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if 400 < mx < 700 and 280 < my < 340:
                state.reset_run()
        return

    if state.phase == GamePhase.GAME_OVER:
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if 400 < mx < 700 and 300 < my < 360:
                state.reset_run()
        return

    if state.phase == GamePhase.PAUSED:
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if 420 < mx < 620 and 200 < my < 250:
                state.phase = GamePhase.PLAYING
            elif 420 < mx < 620 and 270 < my < 320:
                state.reset_run()
            elif 420 < mx < 620 and 340 < my < 390:
                state.phase = GamePhase.MENU
        return

    if state.phase == GamePhase.HELI_SETUP and event.type == pygame.MOUSEBUTTONDOWN:
        if event.button == 1 and state.heli_setup_tower:
            from game.combat import setup_heli

            cx = max(40, min(mx, MAP_PIXEL_W - 40))
            cy = max(40, min(my, MAP_PIXEL_H - 40))
            pad = state.heli_setup_tower
            radius = ((cx - pad.pos[0]) ** 2 + (cy - pad.pos[1]) ** 2) ** 0.5
            setup_heli(pad, (cx, cy), radius)
            state.phase = GamePhase.PLAYING
            state.heli_setup_tower = None
            state.shop_selection = None
        return

    if state.phase != GamePhase.PLAYING:
        return

    if event.type == pygame.MOUSEBUTTONDOWN:
        if event.button == 3:
            t = tower_at_screen(state, mx, my)
            state.placed_selection = t
        elif event.button == 1:
            if state.shop_selection:
                cell = screen_to_grid(mx, my)
                if cell:
                    _try_place(state, cell[0], cell[1], audio)


def update_hover(state, mx, my):
    state.hover_cell = screen_to_grid(mx, my)
    state.hover_tower = tower_at_screen(state, mx, my) if state.hover_cell else None


def _try_place(state, col, row, audio):
    tid = state.shop_selection
    if not tid or tid not in TOWER_TYPES:
        return
    if not can_place(state, col, row):
        return
    cost = TOWER_TYPES[tid]["base_cost"]
    if state.gold < cost:
        return

    from game.entities.tower_base import Tower

    state.gold -= cost
    tower = Tower(tid, col, row, cost)
    state.towers.append(tower)
    state.tower_cells.add((col, row))
    state.recompute_path()
    for e in state.enemies:
        e.request_repath(state.global_path)
    if audio:
        audio.play("place")

    if tid == 9:
        state.phase = GamePhase.HELI_SETUP
        state.heli_setup_tower = tower
    else:
        state.shop_selection = None


def _try_upgrade(state, audio):
    t = state.placed_selection
    if not t or not t.can_upgrade():
        return
    cost = t.upgrade_cost()
    if state.gold < cost:
        return
    state.gold -= cost
    t.upgrade()
    if audio:
        audio.play("upgrade")


def _try_sell(state):
    t = state.placed_selection
    if not t:
        return
    state.gold += t.sell_value()
    state.towers.remove(t)
    state.tower_cells.discard((t.col, t.row))
    state.placed_selection = None
    state.recompute_path()
    for e in state.enemies:
        e.request_repath(state.global_path)
